### PR TITLE
Markus/profile field validation

### PR DESCRIPTION
**Pull request by kuosmark**
# Description

- Add better validation for mentor fields, so that it matches the API.

Fixes [Update Profile page field validation to match the API](https://trello.com/c/i9mE16Ry/986-update-profile-page-field-validation-to-match-the-api)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unittest
- [ ] Cypress e2e -tests

# Caveats

The password field has its own ticket and needs more work: [Frontend Registration page password complexity validation](https://trello.com/c/1Rqe9lvG/703-frontend-registration-page-password-complexity-validation)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by rottabonus — APPROVED** [review]
Minor changes could be nice.
Overall I think that this is good enough. 
We don't have that many forms, so maybe this manual checking is fine.

We do have io-ts (the lib we create codecs) though, and it would not have been difficult to create validators with that.

Any ways its fine

File: src/features/ProfilePage/components/PublicInfo/index.tsx [hunk 1]
```diff
@@ -66,15 +70,44 @@ const PublicInfo = ({ isMobile = false }: Props) => {
...
+      ? null
+      : t('public.mentor.birthYear.invalidError');
+
+  const getDisplayNameError = (): string | null => {
+    if (isDisplayNameTooLong(localData.display_name))
+      return t('public.mentor.displayName.tooLongError');
+    if (isDisplayNameTooShort(localData.display_name))
+      return t('public.mentor.displayName.tooShortError');
+    return null;
+  };
```
[the remark this answers is not in the document]
Comment (kuosmark) [reply src/features/ProfilePage/components/PublicInfo/index.tsx:86]: Done!

File: src/features/ProfilePage/components/PublicInfo/index.tsx [hunk 2]
```diff
@@ -66,15 +70,44 @@ const PublicInfo = ({ isMobile = false }: Props) => {
...
+      : null;
 
-  const isDiscardingDisabled = !isDirty || isLoading;
   const isSavingDisabled =
-    !isDirty || isBirthYearInvalid || isDisplayNameTooShort || isLoading;
+    !!getBirthYearError() ||
+    !!getDisplayNameError() ||
+    !!getRegionError() ||
+    !!getStatusMessageError() ||
+    !!getStoryError() ||
```
[the remark this answers is not in the document]
Comment (kuosmark) [reply src/features/ProfilePage/components/PublicInfo/index.tsx:108]: Yes that's true! I'll change it!

[the remark this answers is not in the document]
Comment (kuosmark) [reply src/features/ProfilePage/components/PublicInfo/index.tsx:108]: I usually do that, don't know why not here.

[the remark this answers is not in the document]
Comment (kuosmark) [reply src/features/ProfilePage/components/PublicInfo/index.tsx]: Done!

File: src/features/ProfilePage/components/PublicInfo/index.tsx [hunk 3]
```diff
@@ -100,18 +135,21 @@ const PublicInfo = ({ isMobile = false }: Props) => {
       <Form>
         <Text>{t('public.mentor.mandatoryNotice')}</Text>
         <Columns
-          isBirthYearInvalid={isBirthYearInvalid}
-          isDisplayNameTooShort={isDisplayNameTooShort}
+          birthYearError={getBirthYearError()}
```
Comment (rottabonus) [inline src/features/ProfilePage/components/PublicInfo/index.tsx:138]: Would it make sense to just create a variable
`const birthYearError = getBirthYearError()`

and use that in the `isSavingDisabled` variable and in the component prop? Now we create one extra function for each.

Even better would be to 

```typescript
const error = getError()

const getError = (): Error | null => {
  if(it_is_this_error){
   return 'this_error'
}
...
return null
}


const errorMap = Record<Error, Message> = {
  this_error: this.error.localization,
  that_error: that.error.localization
}

const isSavingDisabled = Boolean(error) || !isDIrty || isLoading

return (
<Component error={error ? errorMap[error] : null} />
)
```

Now we have a lot of extra props.

**Review by rottabonus — APPROVED** [review]
A new glance on this PR and I would rather refactor a bit.
I think here we have too much boilerplate / extra stuff that does not scale well if we want to refactor error types in the future or so.

I'll approve any way just because time is on the card. Even if I'd like to focus on quality

**Pull request merged**